### PR TITLE
Mocking via inlining 

### PR DESCRIPTION
**Pull request by oleksiyp**
This is second inlining implementation. 
It fully moves from Javassist to ByteBuddy.

Main things achieved:
- use both subclassing and inlining at any time, so abstract class with final method works fine.
  previous implementation had to choose either one or another and fail to handle all cases well
- static method interception
- test coverage for ProxyMaker

### CODE REVIEW

**Comment by oleksiyp** [comment]
Hi @raphw . Can you please review ByteBuddy usage in my mocking library for Kotlin. Thank you in advance.

File: agent/src/main/java/io/mockk/proxy/MockKInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        Class[] cls = classes.toArray(new Class[classes.size()]);
+        try {
+            instrumentation.retransformClasses(cls);
+            return true;
+        } catch (UnmodifiableClassException e) {
+            return false;
+        }
+    }
+
+    public boolean inject(Class<?> clazz) {
```
Comment (oleksiyp) [inline agent/src/main/java/io/mockk/proxy/MockKInstrumentation.java]: check if can delete it?

File: agent/src/main/java/io/mockk/proxy/MockKInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                                            .and(not(isDefaultFinalizer()))
+                                            .and(not(isPackagePrivateJavaMethods()))))
+                    .visit(Advice.withCustomMapping()
+                            .bind(MockKProxyAdviceId.class, staticAdvice.getId())
+                            .to(MockKStaticProxyAdvice.class).on(
+                                    ElementMatchers.<MethodDescription>isStatic().and(not(isTypeInitializer())).and(not(isConstructor()))
+                                            .and(not(isPackagePrivateJavaMethods()))))
+                    .make();
+
+            unloaded.saveIn(new File("."));
```
Comment (oleksiyp) [inline agent/src/main/java/io/mockk/proxy/MockKInstrumentation.java]: remove it

File: agent/src/main/java/io/mockk/proxy/MockKInstrumentationLoader.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import java.io.InputStream;
+import java.lang.instrument.Instrumentation;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.jar.JarOutputStream;
+
+public class MockKInstrumentationLoader {
+    public static final MockKInstrumentationLoader INSTANCE = new MockKInstrumentationLoader();
+
+    public static final String PKG = "io.mockk.proxy.";
```
Comment (oleksiyp) [inline agent/src/main/java/io/mockk/proxy/MockKInstrumentationLoader.java]: private

File: agent/src/main/java/io/mockk/proxy/MockKInstrumentationLoader.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+            return boot;
+        } catch (IOException ex) {
+            return null;
+        }
+    }
+
+    private boolean addClass(JarOutputStream out, String source) throws IOException {
+        source = source.replace('.', '/');
+
+        InputStream inputStream = MockKInstrumentationLoader.class.getClassLoader().getResourceAsStream(source + ".clazz");
```
Comment (oleksiyp) [inline agent/src/main/java/io/mockk/proxy/MockKInstrumentationLoader.java]: add file rename to gradle.build

File: agent/src/main/java/io/mockk/proxy/MockKProxyMaker.java [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+            if (!transformed) {
+                warnOnFinalMethods(clazz);
+            }
+        } else {
+
+            if (!transformed) {
+                throw new MockKAgentException("Failed to create proxy for " + clazz + ".\n" +
+                        "Instrumentation is not available and class is final.\n" +
+                        "Add -javaagent option to your JVM with MockK Java Agent");
```
Comment (oleksiyp) [inline agent/src/main/java/io/mockk/proxy/MockKProxyMaker.java]: better message

File: mockk/src/main/kotlin/io/mockk/impl/CallRecording.kt [hunk 6]
```diff
@@ -216,21 +231,39 @@ internal class CallRecorderImpl(private val gw: MockKGatewayImpl) : CallRecorder
     override fun hintNextReturnType(cls: KClass<*>, n: Int) {
         childTypes[n] = cls
     }
+
+    override fun estimateCallRounds(): Int {
+        return signedCalls
+                .flatMap { it.invocation.args }
+                .filterNotNull()
+                .map {
+                    when (it::class) {
```
Comment (oleksiyp) [inline mockk/src/main/kotlin/io/mockk/impl/CallRecording.kt]: add coment

File: agent/src/main/java/io/mockk/proxy/MockKProxyMaker.java [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+    public void staticUnProxy(Class<?> clazz) {
+        MockKInvocationHandler.HANDLERS.remove(clazz);
+    }
+
+    private static Junction<MethodDescription> any() {
+        return ElementMatchers.any();
+    }
+
+    private List<Class<?>> getAllSuperclasses(Class<?> clazz) {
```
Comment (oleksiyp) [inline agent/src/main/java/io/mockk/proxy/MockKProxyMaker.java]: and interfaces

File: dsl/src/main/kotlin/io/mockk/API.kt [hunk 8]
```diff
@@ -547,11 +579,11 @@ data class MethodDescription(val name: String,
...
  * Mock invocation
  */
-data class Invocation(val self: MockK,
+data class Invocation(val self: Any,
                       val method: MethodDescription,
-                      val superMethod: MethodDescription?,
                       val args: List<Any?>,
-                      val timestamp: Long) {
+                      val timestamp: Long,
+                      val realCall: Any?) {
```
Comment (oleksiyp) [inline dsl/src/main/kotlin/io/mockk/API.kt]: originalCall

**Comment by raphw** [comment]
Things move fast here but looks good. From the code, I assume you are taking Mockito as an example? I think its the best source!

**Comment by oleksiyp** [comment]
Yes, Mockito was taken as an example, but I throw out everything I do not understand or need. So maybe something important but very sophisticated is missing. Eventually would be resolved of course.

**Pull request merged**